src/renderer/pipeline/background.rs

```diff
@@ -0,0 +1,125 @@
+use super::Camera;
+use crate::renderer::QuadVertex;
+use bytemuck::{cast_slice, Pod, Zeroable};
+use std::mem::size_of;
+use std::ops::Range;
+use wgpu::*;
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Pod, Zeroable)]
+pub struct BackgroundFragment {
+    pub position: [f32; 2],
+    pub width: f32,
+    pub color: [f32; 4],
+}
+
+impl BackgroundFragment {
+    const ATTRIBS: [VertexAttribute; 3] =
+        vertex_attr_array![1 => Float32x2, 2 => Float32, 3 => Float32x4];
+
+    fn desc<'a>() -> VertexBufferLayout<'a> {
+        VertexBufferLayout {
+            array_stride: size_of::<Self>() as BufferAddress,
+            step_mode: VertexStepMode::Instance,
+            attributes: &Self::ATTRIBS,
+        }
+    }
+}
+
+fn create_fragment_buffer(device: &Device, size: BufferAddress) -> Buffer {
+    device.create_buffer(&BufferDescriptor {
+        label: Some("Background Instance Buffer"),
+        size,
+        usage: BufferUsages::VERTEX | BufferUsages::COPY_DST,
+        mapped_at_creation: false,
+    })
+}
+
+fn create_pipeline(
+    device: &Device,
+    surface_config: &SurfaceConfiguration,
+    camera: &Camera,
+) -> RenderPipeline {
+    let shader = device.create_shader_module(ShaderModuleDescriptor {
+        label: Some("Background Shader"),
+        source: ShaderSource::Wgsl(include_str!("../shaders/background.wgsl").into()),
+    });
+
+    let pipeline_layout = device.create_pipeline_layout(&PipelineLayoutDescriptor {
+        label: Some("Background Pipeline Layout"),
+        bind_group_layouts: &[&camera.bind_group_layout],
+        push_constant_ranges: &[],
+    });
+
+    device.create_render_pipeline(&RenderPipelineDescriptor {
```

Comment:
```
Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `Background Pipeline`
    Internal error in VERTEX shader: Metal: program_source:32:19: error: duplicate member 'position'
    metal::float2 position [[attribute(1)]];
                  ^
program_source:31:19: note: previous declaration is here
    metal::float2 position [[attribute(0)]];
                  ^
program_source:41:3: error: type 'vs_mainInput' is not valid for attribute 'stage_in'
  vs_mainInput varyings [[stage_in]]
```

It looks similiar to https://github.com/gfx-rs/naga/issues/1490, which has been open for a year with no response. The same issue exists in glyph pipeline.

Reply:
The two positions are in different structs in the `glyph.wgsl` shader. But it looks like it does not create a proper namespace for the struct on metal. I guess the workaround is to make sure the attribute names are unique, even if they are in different structs.